### ISSUE TITLE
[FEATURE] Add support for Azure Managed Identity authentication

### ISSUE THREAD

**Issue by alexott** [databricks/databricks-sdk-py#164]
the `azure_use_msi` is documented in README, but marked as implemented. Adding support for managed identity could simplify deployment of Python-based applications

**Comment by alexott** [comment]
Implementation that could be adopted: https://github.com/alexott/databricks-playground/blob/main/dbsql-with-aad-token/dbsql-with-aad-token-mi.py#L26

**Comment by AndreiCalin24** [comment]
Any updates? Are there plans to support Azure MI?

**Comment by Fjurg** [comment]
I really want this feature in order to not have to think about yet another client secret

**Comment by drdamour** [comment]
not sure if it's already been thought about and decided as a bad option, but in case it's just you aren't aware of it, this should be as straightforward as taking a dependency on https://learn.microsoft.com/en-us/python/api/overview/azure/identity-readme?view=azure-python and wrapping an instance of DefaultAzureCredential (https://learn.microsoft.com/en-us/python/api/azure-identity/azure.identity.defaultazurecredential?view=azure-python)  up in the ClientCredentials implementation https://github.com/databricks/databricks-sdk-py/blob/cc7c2365afe5443f18eb80233f3de6d2e4eb41ea/databricks/sdk/oauth.py#L697C7-L697C24

you could get fancier and switch between the other credential types from azure identity python lib...but DefaultAzureCredential does the right thing 99% of the time IME and gives a LOT of options to configure it from environment variables including supporting client credentials flow (aka 2 legged) and falls down to azure cli/powershell avail sessions which looks like your sdk likes.

the one it doesn't do is the 3 legged web browser flow (the dotnet azure identity does...but not the python i guess).  you can setup your own chain that would do the browser stuff too that you support by leveraging https://learn.microsoft.com/en-us/python/api/azure-identity/azure.identity.interactivebrowsercredential?view=azure-python 

it'd be even slicker if you gave some way to provide any azure.core.credentials.TokenCredential type instance to plug in advanced scenarios.

fwiw DefaultAzureCredential  really should have been named DefaultAzureCredentialTokenProvider cause it's really a way to get Tokens with get_token https://learn.microsoft.com/en-us/python/api/azure-identity/azure.identity.defaultazurecredential?view=azure-python#azure-identity-defaultazurecredential-get-token which includes an access token and you could then wrap that up in your token types used by your sdk.

maybe you don't want to take on the identity lib dependency tree...and i get that. if that's the case adding a way for me to pass in the instance of ClientCredentials would be enough...we can build the azure identity plugin lib easy enough...we just can't contirbute to oss projects unfortunately.

**Comment by alexott** [comment]
@drdamour adding a new library will add ~10 additional libraries that are needed only on one platform. It simply should be implemented by someone...

**Comment by drdamour** [comment]
> [@drdamour](https://github.com/drdamour) adding a new library will add ~10 additional libraries that are needed only on one platform. It simply should be implemented by someone...

yeah i hear you, i hate bringing in huge dependency chains...but fwiw azure identity is a very versatile lib and i've found it to be worth the pain.

> maybe you don't want to take on the identity lib dependency tree...and i get that. if that's the case adding a way for me to pass in the instance of ClientCredentials would be enough

any chance of supporting this kinda suggestion so if my team wants to use azure identity you'd give us a means to do that?

**Comment by alexott** [comment]
Technically, you can plug in your own authenticator. I did this for databricks-sql-connector: https://github.com/alexott/databricks-playground/blob/main/dbsql-with-aad-token/dbsql-with-aad-token-mi.py#L26

**Comment by drdamour** [comment]
> Technically, you can plug in your own authenticator. I did this for databricks-sql-connector: https://github.com/alexott/databricks-playground/blob/main/dbsql-with-aad-token/dbsql-with-aad-token-mi.py#L26

hm...is the equivalent of that for the WorkspaceClient just passing in our own Azure bearer tokens for the workspace scope? 

`w = WorkspaceClient(host=input('Databricks Workspace URL: '), token=DefaultAzureCredential.get_token().access_token)`

i was too focused on the azure section of the docs to notice that very raw approach.  In such case...is there any way to hook into the refresh capabilities or do i gotta hope i do everything in the lifespan of that token?

**Comment by ghjklw** [comment]
The implementation for `databricks-sql-connector`, if I'm not mistaken, doesn't work for all Azure services because the managed identity endpoint is hard coded to the one used by VMs (see documentation for app services and Azure functions as an example: https://learn.microsoft.com/en-us/azure/app-service/overview-managed-identity)

Even though this particular issue can easily be fixed within `databricks-sql-connector`, this outlines the point that maintaining compatibility with all these authentications mechanisms is a huge task, and the library will always lag behind the flexibility offered by Azure's libraries. The situation can become even more challenging when Databricks uses environment variables that are also used by Azure's librairies (e.g. `ARM_*`) but can be interpreted in a different way. This can be especially challenging over time as fixing such inconsistencies is likely to be a breaking change.

I share @drdamour suggestion that it would be really helpful to: 
- enable the use of `azure.identity`. This could be done through an "extra" dependency group to avoid adding it as a dependency for those that don't need it.
- and/or to pass a `TokenProvider` as credential. Azure `TokenProvider` is very easy to define as a Python protocol to avoid any dependency on the `azure-identity` library. This would be more user-friendly than passing the token itself and would be an easy solution to token expiration problems.

**Comment by alexott** [comment]
@drdamour @ghjklw  instead of passing host & token, you can pass `credential_provider`: https://github.com/databricks/databricks-sdk-py/blob/main/databricks/sdk/__init__.py#L170

**Comment by drdamour** [comment]
> [@drdamour](https://github.com/drdamour) [@ghjklw](https://github.com/ghjklw) instead of passing host & token, you can pass `credential_provider`: https://github.com/databricks/databricks-sdk-py/blob/main/databricks/sdk/__init__.py#L170

cool, we'll try it out.  IMO this should be mentioned in the auth docs

**Comment by alexott** [comment]
Here is an example how you can wrap `azure-identity` to use with Python SDK: https://github.com/alexott/databricks-playground/tree/main/azure-mi-in-py-sdk

**Comment by ghjklw** [comment]
Thanks @alexott, that's really useful!

Couldn't a sightly more generic version without any direct dependency on `azure-identity` and `azure-core`, like [this (untested) attempt](https://github.com/ghjklw/databricks-playground/commit/a13c6f671d2af6544479face181e60ef229afe50) make sense to include in `databricks-sdk-py`?

The code snippet you shared with us provides a good solution for users of databricks sdk library directly. I'd like to understand if there could also be a user friendly option for users of other tools (thinking here of `dbt-databricks`, see databricks/dbt-databricks#957) to leverage these additional authentication types when you typically don't directly pass through the `Config` object. 

Of course, each of these tools can also maintain their own logic and create the `credentials_strategy`, but that seems both counter-productive, and a recipe for fragmentation/inconsistent user experience.

Just to explore additional options to balance the convenience/ease-of-use and dependency challenge:

1. What about adding a module `databricks.sdk.azure` which could provide this credential strategy + some options to use it through `Config` object. This module would only ever be imported if `Config.auth_type` is set to one of the new Azure authentication strategies, and `azure-identity` could then be defined as an optional dependency to install when using one of these authentication types?
2. Alternatively, is there a mechanism for registering additional credential strategies that can then be leveraged through config options? If so, would it be possible to build a lightweight independent package/plugin that provides these additional credential strategies? This could for example be something like `DefaultCredentials.register_credential_strategy(strategy: CredentialsStrategy)`.
For it to really make sense though, there should also be a mechanism to extend the `Config` attributes which seems more challenging to implement.

By the way, I'm happy to contribute to either `databricks-sdk-py`, `databricks-sql-connector` or `dbt-databricks` such a functionality, I'm just unclear on the best way to achieve it 😉

**Comment by alexott** [comment]
@ghjklw I'm not a part of the development team, so I can't comment on that :-)

**Comment by ghjklw** [comment]
@alexott, my bad, I assumed you were since I noticed that you work at Databricks 😉

Do you happen to know who could provide us with some guidance?

**Comment by alexott** [comment]
@mgyucht ^
